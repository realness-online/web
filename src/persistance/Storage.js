/** @typedef {import('@/types').Item_Id} Item_Id */

// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import Local from '@/persistance/Local'
import Large from '@/persistance/Large'
import Cloud from '@/persistance/Cloud'
import Paged from '@/persistance/Paged'
import { current_user, upload } from '@/use/serverless'
import { as_type, as_filename } from '@/utils/itemid'
import { get } from 'idb-keyval'
import { prepare_upload_html } from '@/utils/upload_processor'

export default class Storage {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    this.id = itemid
    this.type = as_type(itemid)
    this.metadata = { contentType: 'text/html' }
  }
}

/** @extends {Storage} */
export class Poster extends Large(Cloud(Storage)) {}

/** @extends {Storage} */
export class Me extends Cloud(Local(Storage)) {
  constructor() {
    super(localStorage.me)
  }
}

/** @extends {Storage} */
export class Relation extends Local(Storage) {
  constructor() {
    super(`${localStorage.me}/relations`)
  }
}

/** @extends {Storage} */
export class Statement extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/statements`)
  }
}

/** @extends {Storage} */
export class Event extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/events`)
  }
}

/** @extends {Storage} */
export class Offline extends Cloud(Storage) {
  async save() {
    const outerHTML = await get(this.id)
    if (outerHTML) await super.save({ outerHTML })
  }
}

/** @extends {Storage} */
export class History extends Cloud(Storage) {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    super(itemid)
  }

  async save(items) {
    // on purpose doesn't call super.save
    if (!items) return false
    if (current_user.value && navigator.onLine) {
      const { compressed, metadata } = await prepare_upload_html(items)
      try {
        await upload(as_filename(this.id), compressed, metadata)
      } catch (e) {
        console.error(e)
        return false
      }
      return true
    }
    return false
  }
}
