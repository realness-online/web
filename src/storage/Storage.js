// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import Item from '@/modules/item'
import Paged from '@/storage/Paged'
import Large from '@/storage/Large'
import Cloud from '@/storage/Cloud'

class Storage {
  constructor(type,
              selector = `[itemprop="/${type}"]`,
              filename = `${type}/index.html`,
              content_type = 'text/html') {
    this.type = type
    this.selector = selector
    this.filename = filename
    this.metadata = { 'contentType': content_type }
  }
  as_list() {
    return Item.get_items(localStorage.getItem(this.selector))
  }
  as_object() {
    return Item.get_first_item(localStorage.getItem(this.selector))
  }
  async save(items = document.querySelector(this.selector)) {
    if (!items) return
    localStorage.setItem(this.selector, items.outerHTML)
  }
}
class Person extends Cloud(Storage) {
  constructor() {
    super('person', '[itemid="person.html"]', 'person.html')
  }
}
class Post extends Paged(Cloud(Storage)) {
  constructor() {
    super('posts')
  }
}
class SVG extends Large(Cloud(Storage)) {}

export default Storage
export const relations_storage = new Storage('relations')
export const person_storage = new Person()
export const posts_storage = new Posts()
export const avatars_storage = new SVG('avatars')
export const posters_storage = new SVG('posters')
