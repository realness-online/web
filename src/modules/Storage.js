import Item from '@/modules/Item'
import * as firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/storage'

class Storage {
  static hydrate(item_as_string) {
    return document.createRange().createContextualFragment(item_as_string)
  }
  constructor(item_type, selector = `[itemtype="/${item_type}"]`,
    filename = `${item_type}.html`) {
    this.item_type = item_type
    this.selector = selector
    this.filename = filename
    this.metadata = {'contentType': 'text/html'}
  }
  from_storage() {
    let storage_string = localStorage.getItem(this.item_type)
    return Storage.hydrate(storage_string)
  }
  as_list() {
    return Item.get_items(this.from_storage())
  }
  as_object() {
    return Item.get_first_item(this.from_storage())
  }
  save() {
    // save the information to local storage and if appropriate to the server
    let items = document.querySelector(this.selector)
    if (!items) { return false }
    items = items.outerHTML
    localStorage.setItem(this.item_type, items)
    if (['person', 'posts'].includes(this.item_type)) {
      this.persist(items)
    }
    return true
  }
  sync() {
    // check if an 'item_type'.html already exist for this user.
    // If so merge it's content whith what's stored locally
    const url = this.get_server_file_url()
    if (['person', 'posts'].includes(this.item_type) && navigator.onLine) {
      firebase.auth().onAuthStateChanged(user => {
        if (user) {
          const doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
          firebase.storage().ref().child(doc_u_path).getDownloadURL().then((url) => {
            fetch(url).then(response => {
              response.text().then(server_text => {
                const server_as_fragment = Storage.hydrate(server_text)
                let server_as_list = Item.get_items(server_as_fragment)
                console.log(server_as_list)
              })
              let items = [...this.from_storage, ...from_server]
              // console.log(items)
            })
          })
        }
      })
    }
  }

  get_server_file_url() {

  }

  persist(doc_u_ment, doc_u_path) {
    firebase.auth().onAuthStateChanged(user => {
      if (user && navigator.onLine) {
        const file = new File([doc_u_ment], this.filename)
        if (!doc_u_path) {
          doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
        }
        firebase.storage().ref()
          .child(doc_u_path)
          .put(file, this.metadata)
          .catch(console.log.bind(console))
      }
    })
  }
}

export default Storage
export const person_storage = new Storage('person')
export const posts_storage = new Storage('posts', '[itemprop=posts]')
export const activity_storage = new Storage('activity', '[itemprop=activity]')
export const relations = new Storage('relations', '[itemprop=relations]')
export const phonebook = new Storage('phonebook')

//  use this for firebase logging
// .then( snapshot => {
//   console.log('Uploaded', snapshot.totalBytes, 'bytes.')
//   console.log(snapshot.metadata)
//   console.log(snapshot.fullPath)
//   console.log('File available at', snapshot.downloadURL)
// })
