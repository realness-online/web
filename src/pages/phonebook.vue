<template>
  <section id="directory" class="page">
    <header>
      <label for="search">
        <input id="search" type="search" placeholder="Search phonebook" autocomplete="off"
          v-model="query"
          v-on:focusout="view_friends_mode"
          v-on:focusin="search_mode">
        <icon name="search"></icon>
      </label>
      <h1>Phonebook</h1>
      <logo-as-link></logo-as-link>
    </header>
    <profile-as-list id="phonebook" :people='phonebook'></profile-as-list>
    <footer>
      <menu>
        <router-link to="/relations">Done</router-link>
      </menu>
    </footer>
  </section>
</template>
<script>
  import Vue from 'vue'
  import logo_as_link from '@/components/logo-as-link'
  import icon from '@/components/icon'
  import profile_as_list from '@/components/profile/as-list'
  import {phonebook_storage} from '@/modules/Storage'
  export default {
    components: {
      'logo-as-link': logo_as_link,
      'profile-as-list': profile_as_list,
      icon
    },
    data() {
      return {
        phonebook: [],
        storage: phonebook_storage,
        searching: false,
        query: ''
      }
    },
    created() {
      this.storage.sync_list().then(people => (this.phonebook = people))
    },
    methods: {
      search_mode(event) {
        this.searching = true
      },
      view_friends_mode(event) {
        this.query = ''
        this.searching = false
      }
    },
    watch: {
      phonebook() {
        Vue.nextTick(() => {
          console.log('saving phonebook')
          this.storage.save()
        })
      }
    }
  }
</script>
<style lang='stylus'>
  @require '../style/variables'
  section#directory
    animation-name: slideInLeft
    position: relative
    min-height: 100vh
    & > header
      &:focus-within > h1
        transition-property: all
        overflow: hidden
        width:0
      & > h1
        color:blue
        vertical-align: top
        margin: 0
        line-height: 1.33
        @media (min-width: max-screen)
          line-height: .66
    & > footer
      position: fixed
      bottom: (base-line * 1.5)
      right: base-line
      & menu > a
        standard-button: red

  label[for=search]
    position: relative
    & > *
      fill: blue
      height: (2 * base-line)
      width: (2 * base-line)
    svg
      position: absolute
      top: 0
      left: 0
      z-index: -2
    input#search
      border-width: 0
      border-radius: (base-line / 2)
      position: relative
      z-index: 2
      transition-delay: 0.15s
      transition-property: all
      &::placeholder
        color: transparent
      &:focus
        transition-delay: 0.15s
        standard-border: blue
        padding: (base-line / 2 )
        width: inherit
        &::placeholder
          color:blue
          transition-duration: 0.75
          transition-property: all
          transition-delay: 0.25s
      &:focus ~ svg
        transition-property: all
        height:0
        width:0
</style>
