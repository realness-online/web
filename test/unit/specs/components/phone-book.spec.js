import Vue from 'vue'
import {mount, shallow} from 'vue-test-utils'
import phonebook from '@/components/phone-book'
import Item from '@/modules/Item'
import Storage from '@/modules/Storage'

const phonebook_as_text = `
  <div id="phonebook">
    <figure itemscope itemtype="/person" itemid='+16282281824'>
      <svg><use itemprop="profile_vector" xlink:href="/static/icons.svg#silhouette"></use></svg>
      <figcaption>
        <p><span itemprop="first_name">Scott</span><span itemprop="last_name">Fryxell</span></p>
        <a itemprop="mobile" data-value="+16282281824">+1 (628) 228-1824</a>
      </figcaption>
    </figure>
    <figure itemscope itemtype="/person" itemid='+12403800385‬'>
      <svg><use itemprop="profile_vector" xlink:href="/static/icons.svg#silhouette"></use></svg>
      <figcaption>
        <p><span itemprop="first_name">Katie</span><span itemprop="last_name">Caffey</span></p>
        <a itemprop="mobile" data-value="+12403800385‬">+1 (240) 380-0385‬</a>
      </figcaption>
    </figure>
  <div>`
const save_spy = jest.fn(() => Promise.resolve('save_spy'))
const update_spy = jest.fn((person) => Promise.resolve(people))
const people = Item.get_items(Storage.hydrate(phonebook_as_text))
describe('@/components/phone-book', () => {
  let wrapper
  afterEach(() => {
    save_spy.mockClear()
    update_spy.mockClear()
  })
  it('render an empty phonebook', () => {
    wrapper = shallow(phonebook)
    expect(wrapper.element).toMatchSnapshot()
  })
  it('render a phonebook with people', () => {
    wrapper = mount(phonebook)
    wrapper.setData({phonebook: people})
    expect(wrapper.element).toMatchSnapshot()
  })
  describe('events', () => {
    let storage_info = {
      phonebook_storage: {
        update: update_spy,
        save: save_spy
      }
    }
    beforeEach(() => {
      Vue.prototype.$bus = new Vue({})
      wrapper = shallow(phonebook)
      wrapper.setData(storage_info)
    })
    it('updates the phonebook when a user signs in', () => {
      wrapper.vm.$bus.$emit('signed-in', people[0])
      expect(update_spy).toBeCalled()
    })
    it('updates the phonebook when a user is saved', () => {
      wrapper.vm.$bus.$emit('person-saved', people[0])
      expect(update_spy).toBeCalled()
    })
  })
})
