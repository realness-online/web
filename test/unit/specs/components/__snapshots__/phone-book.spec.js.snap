// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/components/phone-book render a phonebook with people 1`] = `
<section
  class="page"
  id="phonebook"
>
  <header>
    <h1>
      Phone book
    </h1>
  </header>
   
  <nav
    class="profile-index"
  >
    <li>
      <figure
        class="profile"
        itemscope="itemscope"
        itemtype="/person"
      >
        <meta
          itemprop="created_at"
        />
         
        <meta
          itemprop="updated_at"
        />
         
        <a
          href="/profiles/+16282281824"
        >
          <svg>
            <use
              itemprop="image"
              xlink:href="#silhouette"
            />
          </svg>
        </a>
         
        <figcaption>
          <p>
            <span
              itemprop="first_name"
            >
              Scott
            </span>
             
            <span
              itemprop="last_name"
            >
              Fryxell
            </span>
          </p>
           
          <a
            data-value="+16282281824"
            href="sms:+1+16282281824"
            itemprop="mobile"
          >
            +1 628 228 1824
          </a>
        </figcaption>
      </figure>
       
      <menu>
        <!---->
         
        <a>
          <svg>
            <use
              xlink:href="#undefined"
            />
          </svg>
        </a>
      </menu>
    </li>
    <li>
      <figure
        class="profile"
        itemscope="itemscope"
        itemtype="/person"
      >
        <meta
          itemprop="created_at"
        />
         
        <meta
          itemprop="updated_at"
        />
         
        <a
          href="/profiles/+12403800385‬"
        >
          <svg>
            <use
              itemprop="image"
              xlink:href="#silhouette"
            />
          </svg>
        </a>
         
        <figcaption>
          <p>
            <span
              itemprop="first_name"
            >
              Katie
            </span>
             
            <span
              itemprop="last_name"
            >
              Caffey
            </span>
          </p>
           
          <a
            data-value="+12403800385‬"
            href="sms:+1+12403800385‬"
            itemprop="mobile"
          >
            +1 240 380 0385
          </a>
        </figcaption>
      </figure>
       
      <menu>
        <!---->
         
        <a>
          <svg>
            <use
              xlink:href="#undefined"
            />
          </svg>
        </a>
      </menu>
    </li>
  </nav>
</section>
`;

exports[`@/components/phone-book render an empty phonebook 1`] = `
<section
  class="page"
  id="phonebook"
>
  <header>
    <h1>
      Phone book
    </h1>
  </header>
   
  <!---->
</section>
`;
